File: app/core/job.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Dict, List
import enum
# from core.task import Task

class JobType(enum.Enum):
    CN_DAILY_SYNC = "cn_daily_sync" # 同步中国股票的日线数据
    HK_DAILY_SYNC = "hk_daily_sync" # 同步香港股票的日线数据
    US_DAILY_SYNC = "us_daily_sync" # 同步美国股票的日线数据

class JobConcurrencyKey(enum.Enum):
    CN_DAILY_SYNC = "cn_daily_sync"
    HK_DAILY_SYNC = "hk_daily_sync"
    US_DAILY_SYNC = "us_daily_sync"

@dataclass
class JobDefinition:
    type: JobType
    concurrency_key: str
    max_concurrency: int
    singleton: bool = False # whether only one instance of this job type can run at the same time across the whole system

JOB_DEFINITIONS = {
    JobType.US_DAILY_SYNC: JobDefinition(
        type=JobType.US_DAILY_SYNC,
        concurrency_key=JobConcurrencyKey.US_DAILY_SYNC.value,
        max_concurrency=1,
        singleton=True
    ),
    JobType.HK_DAILY_SYNC: JobDefinition(
        type=JobType.HK_DAILY_SYNC,
        concurrency_key=JobConcurrencyKey.HK_DAILY_SYNC.value,
        max_concurrency=1,
        singleton=True
    ),
    JobType.CN_DAILY_SYNC: JobDefinition(
        type=JobType.CN_DAILY_SYNC,
        concurrency_key=JobConcurrencyKey.CN_DAILY_SYNC.value,
        max_concurrency=1,
        singleton=True
    ),
}
class JobStatus(enum.Enum):
    CREATED = "CREATED" # indicates that the job is waiting to be executed
    QUEUED = "QUEUED" # indicates the job is added into quequ and be waiting for execution
    RUNNING = "RUNNING" # indicates that the job is currently being executed
    SUCCESS = "SUCCESS" # indicates that the job has completed successfully
    FAILED = "FAILED" # indicates that the job has completed with a failure

@dataclass
class Job:
    id: str
    type: JobType
    task_id: str | None = None  # task object reference
    task: Task | None = None  # task object reference
    status: JobStatus = JobStatus.CREATED
    params: Dict = field(default_factory=dict)
    depends_on: List[str] = field(default_factory=list)
    
    retry_count: int = 0
    retries: int = 3

    execute_time: str = ""
    stop_time: str = ""
    message: str = ""
    error: str = ""

    def get_id(self):
        return self.id

    def update_status(self, new_status: JobStatus) -> bool:
        self.status = new_status
        # Here you can also add code to log the status change or perform other actions as needed
        if self.task:
            return self.task.update_status_based_on_jobs() 
        else:
            return False

File: quant-mantis/src/modules/task/types/job.ts
// ==============================
// 枚举定义
// ==============================
export enum JobType {
    CN_DAILY_SYNC = "cn_daily_sync",
    HK_DAILY_SYNC = "hk_daily_sync",
    US_DAILY_SYNC = "us_daily_sync",
    PYTHON_SCRIPT = "python_script",
}

export enum JobConcurrencyKey {
    CN_DAILY_SYNC = "cn_daily_sync",
    HK_DAILY_SYNC = "hk_daily_sync",
    US_DAILY_SYNC = "us_daily_sync",
}

export enum JobStatus {
    CREATED = "CREATED",
    QUEUED = "QUEUED",
    RUNNING = "RUNNING",
    SUCCESS = "SUCCESS",
    FAILED = "FAILED",
}

// ==============================
// JobDefinition
// ==============================
export class JobDefinition {
    type: JobType;
    concurrency_key: string;
    max_concurrency: number;
    singleton: boolean;

    constructor(
        type: JobType,
        concurrency_key: string,
        max_concurrency: number,
        singleton = false
    ) {
        this.type = type;
        this.concurrency_key = concurrency_key;
        this.max_concurrency = max_concurrency;
        this.singleton = singleton;
    }
}

// ==============================
// 全局任务定义（对应 Python JOB_DEFINITIONS）
// ==============================
export const JOB_DEFINITIONS: Record<JobType, JobDefinition> = {
    [JobType.US_DAILY_SYNC]: new JobDefinition(
        JobType.US_DAILY_SYNC,
        JobConcurrencyKey.US_DAILY_SYNC,
        1,
        true
    ),
    [JobType.HK_DAILY_SYNC]: new JobDefinition(
        JobType.HK_DAILY_SYNC,
        JobConcurrencyKey.HK_DAILY_SYNC,
        1,
        true
    ),
    [JobType.CN_DAILY_SYNC]: new JobDefinition(
        JobType.CN_DAILY_SYNC,
        JobConcurrencyKey.CN_DAILY_SYNC,
        1,
        true
    ),
    [JobType.PYTHON_SCRIPT]: new JobDefinition(
        JobType.PYTHON_SCRIPT,
        "",
        5,
        false
    ),
};

// ==============================
// Job 实体类
// ==============================

// 解决循环引用：先声明 Task
export interface Task {
    id: string;
    status: any;
    jobs: Job[];
}

export class Job {
    id: string;
    type: JobType;
    task_id: string | null = null;
    task: Task | null = null;
    status: JobStatus = JobStatus.CREATED;
    params: Record<string, any> = {};
    depends_on: string[] = [];

    retry_count = 0;
    retries = 3;

    execute_time = "";
    stop_time = "";
    message = "";
    error = "";
    data_source = "";

    constructor(id: string, type: JobType) {
        this.id = id;
        this.type = type;
    }

    // 对应 Python get_id(self)
    getId(): string {
        return this.id;
    }

    // 对应 Python update_status
    updateStatus(newStatus: JobStatus): JobStatus {
        const oldStatus = this.status;
        this.status = newStatus;
        return oldStatus;
    }
}